umbraco/Umbraco-CMS
# Item tracking improvements

## Details
This PR's purpose is to begin the process of integrating the functionality of the Nexu package directly into Umbraco 9.
The whole feature request can be found [here](https://github.com/umbraco/Umbraco-CMS/issues/8679).
@dawoe had already worked on 5/8 of the items on the _TODO List_ for v8, and this PR migrates those to v9. 

More specifically:
- Tracked references on Content items;
- Even if a parent item is not used, indicate on the info tab that one or more of the underlying items (descendants) is being used;
- Warn an editor when deleting a Media item that is being used;
- Warn an editor when deleting a Content item that is being used;
- Warn an editor when unpublishing a Content item that is being used.

</br>

This PR also adds the last 3 features from the list, related to warning the user when performing listview bulk actions:
- [Warn an editor when deleting a Media item that is being used, using the listview bulk delete action](https://github.com/umbraco/Umbraco-CMS/issues/8696);
- [Warn an editor when deleting a Content item that is being used, using the listview bulk delete action](https://github.com/umbraco/Umbraco-CMS/issues/8697);
- [Warn an editor when unpublishing a Content item that is being, used using the listview bulk unpublish action](https://github.com/umbraco/Umbraco-CMS/issues/8698).

## Tests
- Have the SK installed
  -  Check if media item references are shown correctly on the info tab
      - ex: _Umbraco Campari Meeting Room_ image has a reference to Home
  - Check if content item references are shown correctly on the info tab
      - ex: _Blog_ node has a reference to Home
  - Check that it is indicated on the info tab when underlying items are used (both for Content and Media)
      - ex: _Design_ folder shows "One or more of this item's descendants is being used in a content item."
  - Check that when you try to **delete** a media item that is in use, you get a warning 
      - ex: Before deleting _Umbraco Campari Meeting Room_ image, you should get a warning
  - Check that when you try to **delete** a content item that is in use you get a warning 
      - ex: Before deleting _Blog_ node, you should get a warning
  - Check that when you try to **unpublish** a content item that is in use, you get a warning 
      - ex: Try unpublishing _Biker Jacket_ node should show a warning
  - Check that when trying to bulk **delete** Media items, you will only be warned about the items used in relations
      - ex: Click a Media folder and select used, unused, or both used and unused items and verify that the **delete** warning will show correctly
  - Check that when trying to bulk **delete** Content items, you will only be warned about the items used in relations
      - ex: Click a Products > Child Items and select used, unused, or both used and unused items and verify that the **delete** warning will show correctly
  - Check that when trying to bulk **unpublish** Content items, you will only be warned about the items used in relations
      - ex: Click a Products > Child Items and select used, unused, or both used and unused items and verify that the **unpublish** warning will show correctly
- Create your own references and verify the above ⬆️ 
- You can also modify some media/member types and reference content or media items to get the full set of references ("Used in Documents", "Used in Members", "Used in Media") displayed in the info tab.
- Experiment with pagination (p.1):
    - For info-tab, set ex: `vm.contentOptions.pageSize = 5` in [umbtrackedreferences.component.js loadContentRelations()](https://github.com/umbraco/Umbraco-CMS/blob/44e86d1351cd5e12ce62fd6108dd5773b29870e3/src/Umbraco.Web.UI.Client/src/common/directives/components/references/umbtrackedreferences.component.js#L100)
```js
        function loadContentRelations() {

            vm.contentOptions.pageSize = 5;

            return trackedReferencesResource.getPagedReferences(vm.id, vm.contentOptions)
                  . . .
        }
```
- Experiment with pagination (p.2):
   - For listview, set ex: `vm.mediaOptions.pageSize = 5;` in [umbtrackedreferenceslistview.component.js checkMediaListViewUsage()](https://github.com/umbraco/Umbraco-CMS/blob/44e86d1351cd5e12ce62fd6108dd5773b29870e3/src/Umbraco.Web.UI.Client/src/common/directives/components/references/umbtrackedreferenceslistview.component.js#L72)


```js
        function checkMediaListViewUsage() {
            var ids = vm.selection.map(s => s.id);
          
            vm.mediaOptions.pageSize = 5;

            return trackedReferencesResource.checkLinkedItems(ids, vm.mediaOptions)
                  . . .
        }
```


---
_Update:_
- **This PR also includes the functionality mentioned in** https://github.com/umbraco/Umbraco-CMS/pull/12039 **and** https://github.com/umbraco/Umbraco-CMS/pull/12047
